opte/src/engine/packet.rs

```diff
@@ -751,19 +753,115 @@ impl Packet<Initialized> {
+
+            // The body starts at a zero offset in segment n. This means we need
+            // to squash all segments prior to n.
+            (n, 0) => n - 1,
+
+            // The body starts at a non-zero offset in segment n. This means we
+            // need to squash all segments up to and including n.
+            (n, _) => n,
+        };
+
+        // If the squash bound is zero, there is nothing left to do here, just
+        // return.
+        if squash_to == 0 {
+            return Ok(Packet {
+                avail: self.avail,
+                source: self.source,
+                // The new packet is taking ownership of the segments.
+                segs: core::mem::take(&mut self.segs),
+                state: Parsed {
+                    len: self.state.len,
+                    hdr_offsets: info.offsets,
+                    meta: info.meta,
+                    flow,
+                    body_csum: info.body_csum,
+                    body,
+                    body_modified: false,
+                },
+            });
+        }
+
+        // Calculate the body offset within the new squashed segment
+        if body.seg_offset != 0 {
+            for s in &self.segs[..squash_to] {
+                body.seg_offset += s.len;
+            }
+        }
+        body.seg_index -= squash_to;
+
+        // Determine how big the message block for the squashed segment needs to
+        // be.
+        let mut new_seg_size = 0;
+        for s in &self.segs[..squash_to + 1] {
+            new_seg_size += s.len;
+        }
+
+        let extra_space = info.extra_hdr_space.unwrap_or(0);
+        let mut mp = allocb(new_seg_size + extra_space);
+        unsafe {
+            (*mp).b_wptr = (*mp).b_wptr.add(extra_space);
+            (*mp).b_rptr = (*mp).b_rptr.add(extra_space);
+            for s in &self.segs[..squash_to + 1] {
+                core::ptr::copy_nonoverlapping(
+                    (*s.mp).b_rptr,
+                    (*mp).b_wptr,
+                    s.len,
+                );
+                (*mp).b_wptr = (*mp).b_wptr.add(s.len);
+            }
+        }
+
+        // Construct a new segment vector, tacking on any remaining segments
+        // after the header segments.
+        let orig_segs = core::mem::take(&mut self.segs);
+        let mut segs = vec![unsafe { PacketSeg::wrap_mblk(mp) }];
+        if squash_to + 1 < orig_segs.len() {
+            segs[0].link(&orig_segs[squash_to + 1]);
+            segs.extend_from_slice(&orig_segs[squash_to + 1..]);
+            #[cfg(any(feature = "std", test))]
+            for s in &orig_segs[..squash_to + 1] {
+                mock_freeb(s.mp);
+            }
+        }
+
+        // Recompute info after squash.
+        self.segs = core::mem::take(&mut segs);
+        let mut rdr = self.get_rdr_mut();
+        let info = match dir {
+            Direction::Out => net.parse_outbound(&mut rdr)?,
+            Direction::In => net.parse_inbound(&mut rdr)?,
```

Comment:
a bit unfortunate to have to re-parse here. why not update the offsets in-place?

Reply:
Changed to update in place.